StreamFlix Content Analytics — Phase 3: KPI Calculations

In [1]:
import pandas as pd
from sqlalchemy import create_engine

password = "mayuri27"
engine = create_engine(f"mysql+pymysql://root:{password}@localhost:3306/streamflix")

## KPI 1: Top Genres by Watch Hours

Identifies which genres drive the most total viewing time.

In [2]:
query = """
SELECT t.primary_genre,
       ROUND(SUM(w.watch_duration_min) / 60.0, 1) AS watch_hours,
       COUNT(*) AS plays
FROM watch_history w
JOIN titles t ON t.title_id = w.title_id
GROUP BY t.primary_genre
ORDER BY watch_hours DESC;
"""
df = pd.read_sql_query(query, engine)
df

,primary_genre,watch_hours,plays
0,Drama,501884.5,95624
1,Comedy,423149.1,84455
2,Action,393171.9,77300
3,Thriller,307024.7,58589
4,Documentary,291216.9,57786
5,Crime,237435.8,44153
6,Romance,214332.3,44210
7,Animation,180093.9,29208
8,Sci-Fi,154332.2,33157
9,Horror,138033.6,28122


## KPI 2: Watch Hours by Origin Country (with Efficiency)

Shows which countries' content performs best, and hours generated per title.

In [3]:
query = """
SELECT t.country,
       ROUND(SUM(w.watch_duration_min) / 60.0, 1) AS watch_hours,
       COUNT(DISTINCT t.title_id) AS titles,
       ROUND(SUM(w.watch_duration_min) / 60.0 / COUNT(DISTINCT t.title_id), 1) AS hours_per_title
FROM watch_history w
JOIN titles t ON t.title_id = w.title_id
GROUP BY t.country
ORDER BY watch_hours DESC;
"""
df = pd.read_sql_query(query, engine)
df

,country,watch_hours,titles,hours_per_title
0,United States,822684.7,2265,363.2
1,India,464337.8,1163,399.3
2,United Kingdom,282946.4,769,367.9
3,South Korea,223476.4,636,351.4
4,Japan,207130.9,556,372.5
5,Brazil,176353.6,447,394.5
6,Mexico,157176.9,441,356.4
7,Spain,143656.4,419,342.9
8,France,115011.8,377,305.1
9,Canada,107073.0,297,360.5


## KPI 3: Engagement by Release Year

Shows how well titles from each release year perform, on average, per title.

In [4]:
query = """
SELECT t.release_year,
       COUNT(DISTINCT t.title_id) AS titles,
       ROUND(SUM(w.watch_duration_min) / 60.0, 1) AS watch_hours,
       ROUND(SUM(w.watch_duration_min) / 60.0 / COUNT(DISTINCT t.title_id), 1) AS hours_per_title
FROM watch_history w
JOIN titles t ON t.title_id = w.title_id
GROUP BY t.release_year
ORDER BY t.release_year;
"""
df = pd.read_sql_query(query, engine)
df

,release_year,titles,watch_hours,hours_per_title
0,1969,66,24762.5,375.2
1,1970,58,16432.8,283.3
2,1971,55,20331.8,369.7
3,1972,59,18942.7,321.1
4,1973,55,17519.0,318.5
5,1974,43,6978.1,162.3
6,1975,68,23483.6,345.3
7,1976,59,18484.2,313.3
8,1977,63,16426.2,260.7
9,1978,54,9206.0,170.5


## KPI 4: Movies vs TV Shows — Hours and Completion

Compares the two content types on total hours and average completion rate.

In [5]:
query = """
SELECT t.type,
       ROUND(SUM(w.watch_duration_min) / 60.0, 1) AS watch_hours,
       COUNT(*) AS plays,
       ROUND(AVG(w.completion_pct), 1) AS avg_completion_pct
FROM watch_history w
JOIN titles t ON t.title_id = w.title_id
GROUP BY t.type
ORDER BY watch_hours DESC;
"""
df = pd.read_sql_query(query, engine)
df

,type,watch_hours,plays,avg_completion_pct
0,TV Show,2882608.9,267670,65.2
1,Movie,450858.8,382330,65.4


## KPI 5: Watch Hours by Language

Ranks markets by language to see where demand is strongest.

In [6]:
query = """
SELECT t.language,
       ROUND(SUM(w.watch_duration_min) / 60.0, 1) AS watch_hours,
       COUNT(*) AS plays
FROM watch_history w
JOIN titles t ON t.title_id = w.title_id
GROUP BY t.language
ORDER BY watch_hours DESC;
"""
df = pd.read_sql_query(query, engine)
df

,language,watch_hours,plays
0,English,1352815.8,264430
1,Hindi,464337.8,89686
2,Spanish,347738.2,69080
3,Korean,223476.4,44162
4,Japanese,207130.9,38554
5,Portuguese,176353.6,33063
6,French,115011.8,28375
7,German,96934.4,20586
8,Turkish,82720.0,14185
9,Indonesian,71049.9,12003


## KPI 6: Completion Rate by Genre

Shows which genres keep viewers engaged all the way through.

In [7]:
query = """
SELECT t.primary_genre,
       ROUND(AVG(w.completion_pct), 1) AS avg_completion_pct,
       SUM(CASE WHEN w.completed THEN 1 ELSE 0 END) AS completed_sessions,
       COUNT(*) AS total_sessions
FROM watch_history w
JOIN titles t ON t.title_id = w.title_id
GROUP BY t.primary_genre
ORDER BY avg_completion_pct DESC;
"""
df = pd.read_sql_query(query, engine)
df

,primary_genre,avg_completion_pct,completed_sessions,total_sessions
0,Musical,66.8,1213.0,7425
1,Animation,66.4,4812.0,29208
2,Stand-Up,66.3,1332.0,7993
3,Fantasy,66.0,4312.0,26407
4,Sci-Fi,65.8,5290.0,33157
5,Mystery,65.6,2754.0,18074
6,Crime,65.5,6791.0,44153
7,Comedy,65.3,12803.0,84455
8,Thriller,65.3,8867.0,58589
9,Horror,65.3,4243.0,28122


## KPI 7: Engagement by Subscription Plan

Shows which plan tier watches the most, on average, per subscriber.

In [8]:
query = """
SELECT s.plan_type,
       COUNT(DISTINCT s.subscriber_id) AS subscribers,
       ROUND(SUM(w.watch_duration_min) / 60.0 / COUNT(DISTINCT s.subscriber_id), 1) AS avg_hours_per_subscriber
FROM watch_history w
JOIN subscribers s ON s.subscriber_id = w.subscriber_id
GROUP BY s.plan_type
ORDER BY avg_hours_per_subscriber DESC;
"""
df = pd.read_sql_query(query, engine)
df

,plan_type,subscribers,avg_hours_per_subscriber
0,Premium,4442,228.8
1,Basic with Ads,4458,226.9
2,Standard,6098,214.2


## KPI 8: Device Usage Share

Shows what percentage of all sessions happen on each device type.

In [9]:
query = """
SELECT w.device,
       COUNT(*) AS sessions,
       ROUND(100.0 * COUNT(*) / (SELECT COUNT(*) FROM watch_history), 1) AS pct_of_sessions
FROM watch_history w
GROUP BY w.device
ORDER BY sessions DESC;
"""
df = pd.read_sql_query(query, engine)
df

,device,sessions,pct_of_sessions
0,Smart TV,221446,34.1
1,Mobile,175605,27.0
2,Laptop,96408,14.8
3,Tablet,65276,10.0
4,Game Console,52322,8.0
5,Streaming Stick,38943,6.0


## KPI 9: Churn Rate and Active Rate

The headline retention metric — what share of subscribers are active vs churned.

In [10]:
query = """
SELECT COUNT(*) AS total_subscribers,
       SUM(CASE WHEN is_active THEN 1 ELSE 0 END) AS active,
       ROUND(100.0 * SUM(CASE WHEN is_active THEN 1 ELSE 0 END) / COUNT(*), 1) AS active_rate_pct,
       ROUND(100.0 * SUM(CASE WHEN NOT is_active THEN 1 ELSE 0 END) / COUNT(*), 1) AS churn_rate_pct
FROM subscribers;
"""
df = pd.read_sql_query(query, engine)
df

,total_subscribers,active,active_rate_pct,churn_rate_pct
0,15000,11199.0,74.7,25.3


## KPI 10: Monthly Recurring Revenue (MRR) and ARPU

Calculates total monthly revenue and average revenue per active subscriber.

In [11]:
query = """
SELECT ROUND(SUM(monthly_price_usd), 2) AS mrr_usd,
       COUNT(*) AS active_subscribers,
       ROUND(SUM(monthly_price_usd) / COUNT(*), 2) AS arpu_usd
FROM subscribers
WHERE is_active = TRUE;
"""
df = pd.read_sql_query(query, engine)
df

,mrr_usd,active_subscribers,arpu_usd
0,175868.51,11199,15.7


## KPI 11: Watchlist Conversion Rate

Measures how often subscribers actually watch what they save to their watchlist.

In [12]:
query = """
SELECT COUNT(*) AS watchlist_entries,
       SUM(CASE WHEN watched THEN 1 ELSE 0 END) AS converted,
       ROUND(100.0 * SUM(CASE WHEN watched THEN 1 ELSE 0 END) / COUNT(*), 1) AS conversion_rate_pct
FROM watchlist;
"""
df = pd.read_sql_query(query, engine)
df

,watchlist_entries,converted,conversion_rate_pct
0,65000,30048.0,46.2


## KPI 12: Content Investment Efficiency

Shows watch hours generated per $1,000 spent on licensing, by genre — a measure of ROI.

In [13]:
query = """
SELECT t.primary_genre,
       ROUND(SUM(t.license_cost_usd) / 1000.0, 0) AS spend_thousands_usd,
       ROUND(SUM(t.total_watch_hours), 0) AS watch_hours,
       ROUND(SUM(t.total_watch_hours) / (SUM(t.license_cost_usd) / 1000.0), 3) AS hours_per_1k_usd
FROM titles t
GROUP BY t.primary_genre
ORDER BY hours_per_1k_usd DESC;
"""
df = pd.read_sql_query(query, engine)
df

,primary_genre,spend_thousands_usd,watch_hours,hours_per_1k_usd
0,Animation,868312.0,180093.0,0.207
1,Kids & Family,582701.0,103119.0,0.177
2,Documentary,1701154.0,291218.0,0.171
3,Action,2312734.0,393172.0,0.170
4,Crime,1456105.0,237436.0,0.163
5,Drama,3204980.0,501885.0,0.157
6,Comedy,2762030.0,423148.0,0.153
7,Fantasy,835817.0,127839.0,0.153
8,Thriller,2034649.0,307025.0,0.151
9,Romance,1437719.0,214332.0,0.149


## Overall Summary

Across the 12 KPIs, StreamFlix shows a 74.7% active subscriber rate (25.3% churn), generating $175,868.51 in Monthly Recurring Revenue at an ARPU of $15.70. Engagement is fairly even across plans — Premium (228.8 hrs/subscriber) and Basic with Ads (226.9 hrs) slightly outpace Standard (214.2 hrs). Drama and TV Shows continue to drive the bulk of watch hours, consistent with Phase 2 findings. Watchlist conversion sits at 46.2%, meaning under half of saved titles are actually watched — a potential area for improved recommendations. On content investment efficiency, Animation delivers the strongest return (0.21 watch hours per $1,000 spent), ahead of Kids & Family and Documentary, suggesting licensing spend in these genres is being used efficiently relative to engagement generated. Together, these KPIs point to a healthy but improvable platform: solid revenue and retention, with clear opportunities in watchlist engagement and content investment targeting.